### ISSUE TITLE
False positive `type mapping mismatch: database can contain [json] but property expects [some covariant type]`

### ISSUE THREAD

**Issue by simPod** [phpstan/phpstan-doctrine#196]
So with #195 I'm now getting:

> Property Entity\A::$field type mapping mismatch: database can contain array|bool|float|int|JsonSerializable|stdClass|string|null but property expects array|null.

for

```php
/**
 * @ORM\Column(type="json", nullable=true)
 *
 * @var array<mixed>|null
 */
private array|null $field = null;
```

I don't see what causes the problem yet. I also failed to reproduce it in tests here.

**Comment by ondrejmirtes** [comment]
This is actually expected. You correctly updated the definition - the database can contain all of these types when the JSON value is decoded to PHP value.

**Comment by simPod** [comment]
Oh daym! Thx.

**Comment by ondrejmirtes** [comment]
Maybe we find the new change more annoying in practice and will decide to revert it. Let's wait for other reports.

**Comment by ondrejmirtes** [comment]
You should probably use json_array instead.

**Comment by simPod** [comment]
That one is deprecated in favour of json IIRC. I'll adjust my private read/write methods for the `json` type fields.

**Comment by benja-M-1** [comment]
Hello, I am struggling to fix the same error in my code:

```php
class User implements UserInterface, PasswordAuthenticatedUserInterface
{
    /**
     * @ORM\Column(type="json")
     */
    private array $roles = [];

    /**
     * @see UserInterface
     */
    public function getRoles(): array
    {
        $roles = $this->roles;
        // guarantee every user at least has ROLE_USER
        $roles[] = 'ROLE_USER';

        return array_unique($roles);
    }

    public function setRoles(array $roles): void
    {
        $this->roles = $roles;
    }
}
```

When I add the `@var array|bool|float|int|JsonSerializable|stdClass|string|null` I get the following errors:
```
PHPDoc tag @var for property App\Entity\User::$roles with type
         array|bool|float|int|JsonSerializable|stdClass|string|null is not subtype of native type
         array.
  36     Property App\Entity\User::$roles type has no value type specified in iterable type array.
         💡 See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  36     Property App\Entity\User::$roles type mapping mismatch: database can contain
         array|bool|float|int|JsonSerializable|stdClass|string|null but property expects array.
```

I don't see how to fix this :/

**Comment by ThomasTr** [comment]
Is there a way to explicitly handle only arrays without disabling the rule?

**Comment by ondrejmirtes** [comment]
I'm reverting the change, it proved to be too cumbersome without benefits. So next release of phpstan-doctrine will not report these errors, thanks.

**Comment by ondrejmirtes** [comment]
Alright, I did something that makes the false-positives go away why still disallowing something that cannot be decoded as JSON: https://github.com/phpstan/phpstan-doctrine/commit/062fc75df9a393dc38a9722e0e8ab7036f5429ae

**Comment by benja-M-1** [comment]
Thank you @ondrejmirtes it works! :)
